packages/hdwallet-tallyho/src/tallyho.ts

```diff
@@ -214,6 +262,53 @@ export class TallyHoHDWallet implements core.HDWallet, core.ETHWallet {
     return chainId === 1;
   }
 
+  private async detectTallyProvider(): Promise<TallyHoEthereumProvider | null> {
+    let handled = false;
+
+    return new Promise((resolve) => {
+      if ((window as Window).ethereum) {
+        // eslint-disable-next-line @typescript-eslint/no-use-before-define
+        handleEthereum();
+      } else {
+        // eslint-disable-next-line @typescript-eslint/no-use-before-define
+        window.addEventListener("ethereum#initialized", handleEthereum, { once: true });
+
+        setTimeout(() => {
+          // eslint-disable-next-line @typescript-eslint/no-use-before-define
+          handleEthereum();
+        }, 3000);
+      }
+
+      function handleEthereum() {
+        if (handled) {
+          return;
+        }
+        handled = true;
+
+        window.removeEventListener("ethereum#initialized", handleEthereum);
+
+        const { ethereum } = window as Window;
+
+        if (ethereum && ethereum.isTally) {
+          resolve(ethereum as unknown as TallyHoEthereumProvider);
+        } else {
+          const message = ethereum ? "Non-TallyHo window.ethereum detected." : "Unable to detect window.ethereum.";
+
+          console.error("hdwallet-tallyho: ", message);
+          resolve(null);
+        }
+      }
+    });
+  }
+
+  public async ethGetChainId(): Promise<number> {
+    // at this point, we know that we're in the context of a valid TallyHo provider
+    const provider: any = await this.detectTallyProvider();
+    // chainId as hex string
+    const chainId: string = await provider.request({ method: "eth_chainId" });
+    return parseInt(chainId, 16);
```

Comment:
I'm not sure what possible failure conditions exist for the async function calls here, but maybe this should be wrapped in a `try/catch` for safety.

Reply:
Good point, I'm not sure about this one.
Since we've narrowed the provider itself to be valid, I would think anything going wrong with calling the Ethereum JSON-RPC provider with such method is really an edge case, what do you think? How would we handle the failed call?

Reply:
The particular way we are using hdwallet in web guarantees that we'll be in the context of a valid TallyHo provider here, but hdwallet is used in other non-shapeshift projects. I'd abandon any assumptions about use case here and account for the cases where either of the await-ed calls fail. In this case, I'd throw an error, log the error to console, and change the return type to `Promise<number | null>`.